README
A color management plugin for compiz 0.9.

Dependencies are only compiz, lcms and libcolord.  The ICC profile is retrieved
using colord or the _ICC_PROFILE atom as set by gnome-color-manager.
Compiz-style window exclusions are supported.

To install use my PPA: https://launchpad.net/~gebner/+archive/compiz-cms
$ sudo add-apt-repository ppa:gebner/compiz-cms
$ sudo apt-get update
$ sudo apt-get install compiz-cms
And enable in CCSM.


debian/changelog
compiz-cms (0.5ubuntu1) utopic; urgency=medium

  * Also build PPA for utopic.

 -- Gabriel Ebner <gebner@gebner.org>  Sat, 14 Jun 2014 13:54:07 +0200

compiz-cms (0.5) saucy; urgency=medium

  * Fix multi-monitor support.

 -- Gabriel Ebner <gebner@gebner.org>  Sat, 14 Jun 2014 13:33:01 +0200

compiz-cms (0.4) saucy; urgency=low

  * Support Mesa3d drivers.
  * Use asynchronous libcolord interface.

 -- Gabriel Ebner <gebner@gebner.org>  Thu, 06 Feb 2014 12:28:25 +0100

compiz-cms (0.3.2ubuntu1) saucy; urgency=low

  * Release for saucy.

 -- Gabriel Ebner <gebner@gebner.org>  Sun, 02 Feb 2014 21:36:55 +0100

compiz-cms (0.3.2) raring; urgency=low

  * Improve alpha handling.

 -- Gabriel Ebner <gebner@gebner.org>  Wed, 19 Jun 2013 18:44:39 +0200

compiz-cms (0.3.1) raring; urgency=low

  * Fix grid calculation inaccuracy.

 -- Gabriel Ebner <gebner@gebner.org>  Sun, 16 Jun 2013 21:53:08 +0200

compiz-cms (0.3-6ubuntu3) raring; urgency=low

  * Remove versioned dependency on compiz.

 -- Gabriel Ebner <gebner@2b7e.org>  Fri, 26 Apr 2013 12:35:20 +0200

compiz-cms (0.3-6ubuntu2) raring; urgency=low

  * Bump dependency on compiz.

 -- Gabriel Ebner <gebner@2b7e.org>  Sun, 31 Mar 2013 14:50:12 +0200

compiz-cms (0.3-6ubuntu1) quantal; urgency=low

  * Bump dependency on compiz.

 -- Gabriel Ebner <gebner@2b7e.org>  Wed, 19 Dec 2012 22:46:05 +0100

compiz-cms (0.3-6) quantal; urgency=low

  * Add build dependency on libboost-dev.

 -- Gabriel Ebner <gebner@2b7e.org>  Sun, 11 Nov 2012 15:43:56 +0100

compiz-cms (0.3-5) quantal; urgency=low

  * Add build dependency on xsltproc.

 -- Gabriel Ebner <gebner@2b7e.org>  Sun, 11 Nov 2012 15:26:11 +0100

compiz-cms (0.3-4) quantal; urgency=low

  *  Fix dependency on compiz.

 -- Gabriel Ebner <gebner@2b7e.org>  Sun, 11 Nov 2012 15:00:40 +0100

compiz-cms (0.3-3) quantal; urgency=low

  * Initial release.

 -- Gabriel Ebner <gebner@2b7e.org>  Sun, 11 Nov 2012 14:35:02 +0100


src/cms.h
/* vim:sts=4 sw=4:
 *
 * Copyright (c) 2006 Darryll Truchan <moppsy@comcast.net>
 *	     (C) 2011 Gabriel Ebner <gebner@2b7e.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 */

#include <core/core.h>
#include <core/pluginclasshandler.h>

#include <composite/composite.h>
#include <opengl/opengl.h>

#include <X11/extensions/Xrandr.h>
#include <boost/ptr_container/ptr_vector.hpp>
#include <memory>

#include <lcms.h>
#include <colord.h>

#include "cms_options.h"

struct CmsLut {
    CmsLut(cmsHPROFILE profile);
    ~CmsLut();

    GLuint texture_id;

    std::string getFragmentShader(bool alpha);

    static CmsLut *fromFile(const char *filename);
    static CmsLut *fromMemory(unsigned char *icc, int len);
};

struct CmsOutput {
    CmsOutput(CompScreen *screen, RROutput output);
    ~CmsOutput();

    CompScreen *screen;
    RROutput output;
    CompRect rect;
    bool connected;
    CmsLut *lut;
    bool hasPerOutputProfile;
    std::string name;

    void setLUT(CmsLut *lut);

    void updateRect();
    void updateLUT();

    CdDevice *device;
    void setDevice(CdDevice *);
    char *getColordProfileFilename();
    static void onDeviceChange(CdDevice *dev, CmsOutput *out);
    gulong onDeviceChangeId;
};

class CmsScreen :
    public PluginClassHandler<CmsScreen, CompScreen>,
    public ScreenInterface,
    public CmsOptions
{
public:

    CmsScreen(CompScreen *);
    virtual ~CmsScreen();

    GLScreen *gScreen;

    boost::ptr_vector<CmsOutput> cmsOutputs;

    Atom _ICC_PROFILE;
    int randrEvent, randrError;

    CdClient *cd_client;
    static void onCdConnectFinish(CdClient *, GAsyncResult *, CmsScreen *);
    template <bool unrefAfterUse> void connectToDevice(CdDevice *);
    static void onCdDeviceAdded(CdClient *, CdDevice *, CmsScreen *);
    static void onCdDeviceRemoved(CdClient *, CdDevice *, CmsScreen *);
    template <bool unrefAfterUse> static void onCdDeviceConnectFinish(CdDevice *, GAsyncResult *, CmsScreen *);
    static void onCdClientFindDeviceByPropertyFinish(CdClient *, GAsyncResult *, CmsScreen *);

    void optionChanged(CompOption *opt, CmsOptions::Options num);
    void handleEvent(XEvent *event);

    void setupOutputs();

    bool hasPerOutputProfiles();
};

class CmsWindow :
    public PluginClassHandler<CmsWindow, CompWindow>,
    public GLWindowInterface
{
public:

    CmsWindow(CompWindow *);
    ~CmsWindow();

    CompWindow      *window;
    CompositeWindow *cWindow;
    GLWindow        *gWindow;

    bool isCms;

    void updateMatch();

    void glDrawTexture(GLTexture *texture, const GLMatrix &transform,
	const GLWindowPaintAttrib &attrib, unsigned int mask);
};

#define CMS_SCREEN(s)							      \
    CmsScreen *cs = CmsScreen::get(s);

#define CMS_WINDOW(w)							      \
    CmsWindow *cw = CmsWindow::get(w);

class CmsPluginVTable :
    public CompPlugin::VTableForScreenAndWindow<CmsScreen, CmsWindow>
{
public:

    bool init();
};


.ycm_extra_conf.py
import os
import ycm_core

# These are the compilation flags that will be used in case there's no
# compilation database set (by default, one is not set).
# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
flags = [
'-xc++',
'-std=c++11', '-stdlib=libc++',
]


compilation_database_folder = 'build'

if os.path.exists( compilation_database_folder ):
  database = ycm_core.CompilationDatabase( compilation_database_folder )
else:
  database = None

SOURCE_EXTENSIONS = [ '.cpp' ]

def DirectoryOfThisScript():
  return os.path.dirname( os.path.abspath( __file__ ) )


def MakeRelativePathsInFlagsAbsolute( flags, working_directory ):
  if not working_directory:
    return list( flags )
  new_flags = []
  make_next_absolute = False
  path_flags = [ '-isystem', '-I', '-iquote', '--sysroot=' ]
  for flag in flags:
    new_flag = flag

    if make_next_absolute:
      make_next_absolute = False
      if not flag.startswith( '/' ):
        new_flag = os.path.join( working_directory, flag )

    for path_flag in path_flags:
      if flag == path_flag:
        make_next_absolute = True
        break

      if flag.startswith( path_flag ):
        path = flag[ len( path_flag ): ]
        new_flag = path_flag + os.path.join( working_directory, path )
        break

    if new_flag:
      new_flags.append( new_flag )
  return new_flags


def IsHeaderFile( filename ):
  extension = os.path.splitext( filename )[ 1 ]
  return extension in [ '.h', '.hxx', '.hpp', '.hh' ]


def GetCompilationInfoForFile( filename ):
  # The compilation_commands.json file generated by CMake does not have entries
  # for header files. So we do our best by asking the db for flags for a
  # corresponding source file, if any. If one exists, the flags for that file
  # should be good enough.
  if IsHeaderFile( filename ):
    basename = os.path.splitext( filename )[ 0 ]
    for extension in SOURCE_EXTENSIONS:
      replacement_file = basename + extension
      if os.path.exists( replacement_file ):
        compilation_info = database.GetCompilationInfoForFile(
          replacement_file )
        if compilation_info.compiler_flags_:
          return compilation_info
    return None
  return database.GetCompilationInfoForFile( filename )


def FlagsForFile( filename, **kwargs ):
  if database:
    # Bear in mind that compilation_info.compiler_flags_ does NOT return a
    # python list, but a "list-like" StringVec object
    compilation_info = GetCompilationInfoForFile( filename )
    if not compilation_info:
      return None

    final_flags = MakeRelativePathsInFlagsAbsolute(
      flags + list(compilation_info.compiler_flags_),
      compilation_info.compiler_working_dir_ )

  else:
    relative_to = DirectoryOfThisScript()
    final_flags = MakeRelativePathsInFlagsAbsolute( flags, relative_to )

  return {
    'flags': final_flags,
    'do_cache': True
  }


src/cms.cpp
/* vim:sts=4 sw=4:
 *
 * Copyright (c) 2006 Darryll Truchan <moppsy@comcast.net>
 *	     (C) 2008 Gerhard Fürnkranz
 *	     (C) 2008 Tomas Carnecky
 *	     (C) 2009-2011 Kai-Uwe Behrmann
 *	     (C) 2011 Gabriel Ebner <gebner@2b7e.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 */

#include "cms.h"
#include <lcms.h>
#include <GL/glext.h>
#include <X11/Xatom.h>
#include <limits.h>

#define compWarning(msg...) \
    compLogMessage("cms", CompLogLevelWarn, msg)

COMPIZ_PLUGIN_20090315 (cms, CmsPluginVTable);

const int GRIDSIZE = 64;
struct memlut {
    GLushort a[GRIDSIZE][GRIDSIZE][GRIDSIZE][3];
};

CmsLut::CmsLut(cmsHPROFILE monitorProfile) : texture_id(0) {
    // populate sampling grid
    std::auto_ptr<memlut> in(new memlut),
			  out(new memlut);

    for (int r = 0; r < GRIDSIZE; r++) {
	for (int g = 0; g < GRIDSIZE; g++) {
	    for (int b = 0; b < GRIDSIZE; b++) {
		in->a[b][g][r][0] = floor((double) r / (GRIDSIZE - 1) * 65535.0 + 0.5);
		in->a[b][g][r][1] = floor((double) g / (GRIDSIZE - 1) * 65535.0 + 0.5);
		in->a[b][g][r][2] = floor((double) b / (GRIDSIZE - 1) * 65535.0 + 0.5);
	    }
	}
    }

    // transform
    cmsHPROFILE inputProfile = cmsCreate_sRGBProfile();
    cmsHTRANSFORM transf = cmsCreateTransform(
	inputProfile, TYPE_RGB_16,
	monitorProfile, TYPE_RGB_16,
	INTENT_PERCEPTUAL,
	cmsFLAGS_NOTPRECALC);

    cmsDoTransform(transf, in.get(), out.get(), GRIDSIZE*GRIDSIZE*GRIDSIZE);

    cmsDeleteTransform(transf);
    cmsCloseProfile(inputProfile);

    // save into a texture
    glGenTextures(1, &texture_id);
    glBindTexture(GL_TEXTURE_3D, texture_id);

    glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_WRAP_S, GL_CLAMP);
    glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_WRAP_T, GL_CLAMP);
    glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_WRAP_R, GL_CLAMP);
    glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_3D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);

    glTexImage3D(GL_TEXTURE_3D, 0, GL_RGB16, GRIDSIZE,GRIDSIZE,GRIDSIZE,
	0, GL_RGB, GL_UNSIGNED_SHORT, out->a);
}

CmsLut::~CmsLut() {
    if (texture_id) {
	glDeleteTextures(1, &texture_id);
	texture_id = 0;
    }
}

std::string CmsLut::getFragmentShader(bool alpha) {
    float scale = (float) (GRIDSIZE - 1) / GRIDSIZE;
    float offset = (float) 1.0 / (2 * GRIDSIZE);
    std::ostringstream fs;
    fs << "uniform sampler3D cms_lut;\n";
    fs << "void cms_fragment() {\n";
    fs << "  vec3 color = gl_FragColor.rgb * vec3("<<scale<<") + vec3("<<offset<<");\n";
    fs << "  color = vec3(texture3D(cms_lut, color.rgb));\n";
    // The display profile is only valid for rgb values that are actually sent
    // to the display.  If alpha < 1, then we are correcting rgb values that
    // are never shown on the screen -- this results in a greenish tint in
    // shadow effects on my monitor.
    // So we do a simple linear interpolation here, giving the corrected values
    // if alpha = 1, and the uncorrected values if alpha = 0.
    fs << "  color = gl_FragColor.a * color + (1.0-gl_FragColor.a) * gl_FragColor.rgb;\n";
    fs << "  gl_FragColor = vec4(color, gl_FragColor.a);\n";
    fs << "}\n";
    return fs.str();
}

CmsLut *CmsLut::fromFile(const char *filename) {
    cmsHPROFILE profile = cmsOpenProfileFromFile(filename, "r");
    if (!profile) return 0;

    CmsLut *lut = new CmsLut(profile);

    cmsCloseProfile(profile);
    
    return lut;
}

CmsLut *CmsLut::fromMemory(unsigned char *icc, int len) {
    cmsHPROFILE profile = cmsOpenProfileFromMem(icc, len);
    if (!profile) return 0;

    CmsLut *lut = new CmsLut(profile);

    cmsCloseProfile(profile);
    
    return lut;
}

CmsOutput::CmsOutput(CompScreen *screen, RROutput output)
	: screen(screen), output(output), rect(), connected(false),
	lut(0), hasPerOutputProfile(true), name(), device(0) {
    updateRect();
}

void CmsOutput::setLUT(CmsLut *newLut) {
    if (lut) {
	delete lut;
    }
    lut = newLut;

    foreach (CompWindow *window, screen->windows()) {
	CompositeWindow::get(window)->addDamage();
    }
}

void CmsOutput::updateRect() {
    XRRScreenResources *res = XRRGetScreenResources(
	screen->dpy(), screen->root());
    if (!res) {
	return;
    }

    XRROutputInfo *outputInfo = XRRGetOutputInfo(
	screen->dpy(), res, output);
    if (!outputInfo) {
	XRRFreeScreenResources(res);
	return;
    }
	
    name = outputInfo->name;
    connected = outputInfo->connection != RR_Disconnected;

    XRRCrtcInfo *crtcInfo = XRRGetCrtcInfo(
	screen->dpy(), res, outputInfo->crtc);
    if (!crtcInfo) {
	XRRFreeOutputInfo(outputInfo);
	XRRFreeScreenResources(res);
	return;
    }

    rect = CompRect(crtcInfo->x, crtcInfo->y,
	crtcInfo->width, crtcInfo->height);

    XRRFreeCrtcInfo(crtcInfo);
    XRRFreeOutputInfo(outputInfo);
    XRRFreeScreenResources(res);
}

void CmsOutput::updateLUT() {
    CmsLut *newLut;
    unsigned char *icc;
    Atom type;
    int format, res;
    unsigned long len, bytes_left;

    // fetch ICC profile from colord
    char *filename = getColordProfileFilename();
    if (filename) {
	newLut = CmsLut::fromFile(filename);
	free(filename);
	if (newLut) {
	    setLUT(newLut);
	    return;
	}
    }

    // fetch ICC profile from output _ICC_PROFILE property
    CMS_SCREEN (screen);

    res = XRRGetOutputProperty(screen->dpy(), output,
	cs->_ICC_PROFILE,
	0, LONG_MAX,
	false,
	false,
	XA_CARDINAL,
	&type,
	&format,
	&len, &bytes_left,
	&icc);
    if (res == Success && len != 0) {
	CmsLut *newLut = CmsLut::fromMemory(icc, len);
	XFree(icc);

	if (newLut) {
	    setLUT(newLut);
	    return;
	}
    }
    
    // fetch ICC profile from root _ICC_PROFILE property
    res = XGetWindowProperty(screen->dpy(), screen->root(),
	cs->_ICC_PROFILE,
	0, LONG_MAX,
	false,
	XA_CARDINAL,
	&type,
	&format,
	&len, &bytes_left,
	&icc);
    if (res == Success && len != 0) {
	CmsLut *newLut = CmsLut::fromMemory(icc, len);
	XFree(icc);

	if (newLut) {
	    hasPerOutputProfile = false;
	    setLUT(newLut);
	    return;
	}
    }

    // no profile found, so remove lut
    setLUT(0);
}

void CmsOutput::setDevice(CdDevice *d) {
    if (device) {
	g_signal_handler_disconnect(device, onDeviceChangeId);
	g_object_unref(device);
    }
    device = d;

    if (device) {
	onDeviceChangeId = g_signal_connect(device, "changed", G_CALLBACK(onDeviceChange), this);
    }
}

void CmsOutput::onDeviceChange(CdDevice *dev, CmsOutput *out) {
    out->updateLUT();
}

char *CmsOutput::getColordProfileFilename() {
    if (!device) return 0;

    CdProfile *profile = cd_device_get_default_profile(device);
    if (!profile) {
	compWarning("cannot find profile for colord device %s (%s)",
	    cd_device_get_object_path(device), name.c_str());
	return 0;
    }

    GError *error = 0;
    if (!cd_profile_connect_sync(profile, 0, &error)) {
	compWarning("cannot connect to profile %s for colord device %s (%s)",
	    cd_profile_get_object_path(profile), cd_device_get_object_path(device), name.c_str());
	g_clear_error(&error);
	g_object_unref(profile);
	return 0;
    }

    const char *filename = cd_profile_get_filename(profile);
    if (!filename) {
	compWarning("cannot get filename for profile %s for colord device %s (%s)",
	    cd_profile_get_object_path(profile), cd_device_get_object_path(device), name.c_str());
	g_object_unref(profile);
	return 0;
    }

    char *copy = strdup(filename);
    g_object_unref(profile);
    return copy;
}

CmsOutput::~CmsOutput() {
    setLUT(0);
    setDevice(0);
}

CmsScreen::CmsScreen(CompScreen *screen) :
	PluginClassHandler<CmsScreen, CompScreen>(screen),
	CmsOptions(),
	gScreen(GLScreen::get(screen)) {
    ScreenInterface::setHandler(screen, false);

    optionSetExcludeMatchNotify(boost::bind(&CmsScreen::optionChanged, this, _1, _2));
    optionSetDecorationsNotify( boost::bind(&CmsScreen::optionChanged, this, _1, _2));

    _ICC_PROFILE = XInternAtom(screen->dpy(), "_ICC_PROFILE", false);
    XRRQueryExtension(screen->dpy(), &randrEvent, &randrError);

    XRRSelectInput(screen->dpy(), screen->root(),
	RROutputPropertyNotifyMask | RROutputChangeNotifyMask);
    screen->handleEventSetEnabled(this, true);

    cd_client = cd_client_new();
    cd_client_connect(cd_client, 0, (GAsyncReadyCallback) onCdConnectFinish, this);

    setupOutputs();
}

void CmsScreen::onCdConnectFinish(CdClient *, GAsyncResult *res, CmsScreen *cs) {
    GError *error = 0;
    if (!cd_client_connect_finish(cs->cd_client, res, &error)) {
	compWarning("cannot connect to colord: %s", error ? error->message : "(null)");
	g_clear_error(&error);
	g_object_unref(cs->cd_client);
	cs->cd_client = 0;
    }

    if (cs->cd_client) {
	g_signal_connect(cs->cd_client, "device-added", G_CALLBACK(onCdDeviceAdded), cs);
	g_signal_connect(cs->cd_client, "device-removed", G_CALLBACK(onCdDeviceRemoved), cs);
    }

    foreach (CmsOutput& output, cs->cmsOutputs) {
        if (!output.connected) continue;

        cd_client_find_device_by_property(
            cs->cd_client, CD_DEVICE_METADATA_XRANDR_NAME, output.name.c_str(),
            0, (GAsyncReadyCallback) onCdClientFindDeviceByPropertyFinish, cs);
    }
}

CmsScreen::~CmsScreen() {
    if (cd_client) g_object_unref(cd_client);
}

void CmsScreen::optionChanged(CompOption *opt, CmsOptions::Options num) {
    switch (num) {
    case CmsOptions::Decorations:
    case CmsOptions::ExcludeMatch: {
	foreach (CompWindow *w, screen->windows()) {
	    CMS_WINDOW (w);
	    cw->updateMatch ();
	}
	break;
    }
    default: break;
    }
}

void CmsScreen::handleEvent(XEvent *event) {
    screen->handleEvent (event);

    if (event->type == PropertyNotify
	    && event->xproperty.window == screen->root()
	    && event->xproperty.atom == _ICC_PROFILE) {
	if (!hasPerOutputProfiles()) {
	    foreach (CmsOutput& output, cmsOutputs) {
		output.updateLUT();
	    }
	}
    } else if (event->type == randrEvent + RRNotify
	    && ((XRRNotifyEvent *) event)->subtype == RRNotify_OutputProperty) {
	XRROutputPropertyNotifyEvent *ev =
	    (XRROutputPropertyNotifyEvent *) event;
	if (ev->property == _ICC_PROFILE) {
	    foreach (CmsOutput& output, cmsOutputs) {
		if (output.output == ev->output) {
		    output.updateRect();
		    output.updateLUT();
		}
	    }
	}
    } else if (event->type == randrEvent + RRNotify
	    && ((XRRNotifyEvent *) event)->subtype == RRNotify_OutputChange) {
	XRROutputChangeNotifyEvent *ev =
	    (XRROutputChangeNotifyEvent *) event;
	
	foreach (CmsOutput& output, cmsOutputs) {
	    if (output.output == ev->output) {
		output.updateRect();
		output.updateLUT();
	    }
	}
    }
}

template <bool unrefAfterUse>
void CmsScreen::connectToDevice(CdDevice *device) {
    cd_device_connect(device, 0, (GAsyncReadyCallback) onCdDeviceConnectFinish<unrefAfterUse>, this);
}

void CmsScreen::onCdDeviceAdded(CdClient *, CdDevice *device, CmsScreen *cs) {
    cs->connectToDevice<false>(device);
}

template <bool unrefAfterUse>
void CmsScreen::onCdDeviceConnectFinish(CdDevice *device, GAsyncResult *res, CmsScreen *cs) {
    GError *error = 0;
    if (!cd_device_connect_finish(device, res, &error)) {
	compWarning("cannot connect to colord device %s: %s",
	    cd_device_get_object_path(device), error ? error->message : "(null)");
	g_clear_error(&error);
	return;
    }

    const char *name = cd_device_get_metadata_item(device, CD_DEVICE_METADATA_XRANDR_NAME);
    foreach (CmsOutput& output, cs->cmsOutputs) {
	if (name != 0 && output.name == name) {
	    g_object_ref(device);
	    output.setDevice(device);
	    return;
	}
    }

    compWarning("could not find output for colord device: %s",
	cd_device_get_object_path(device));

    if (unrefAfterUse) {
        g_object_unref(device);
    }
}

void CmsScreen::onCdDeviceRemoved(CdClient *, CdDevice *device, CmsScreen *cs) {
    foreach (CmsOutput& output, cs->cmsOutputs) {
	if (output.device) {
	    if (!strcmp(cd_device_get_object_path(device),
		    cd_device_get_object_path(output.device))) {
		output.setDevice(0);
		return;
	    }
	}
    }

    compWarning("could not find output for removed colord device: %s",
	cd_device_get_object_path(device));
}

void CmsScreen::setupOutputs() {
    XRRScreenResources *res =
	XRRGetScreenResources(screen->dpy(), screen->root());
    if (!res) return;

    for (int i = 0; i < res->noutput; i++) {
	CmsOutput *output = new CmsOutput(screen, res->outputs[i]);
	output->updateLUT();
	cmsOutputs.push_back(output);
    }

    XRRFreeScreenResources(res);
}

void CmsScreen::onCdClientFindDeviceByPropertyFinish(CdClient *client,
        GAsyncResult *res, CmsScreen *cs) {
    GError *error = 0;
    CdDevice *device = cd_client_find_device_by_property_finish(client, res, &error);
    if (!device) {
	g_clear_error(&error);
	return;
    }

    cs->connectToDevice<true>(device);
}

bool CmsScreen::hasPerOutputProfiles() {
    foreach (CmsOutput& output, cmsOutputs) {
	if (!output.hasPerOutputProfile) {
	    return false;
	}
    }
    return true;
}

CmsWindow::CmsWindow(CompWindow *window) :
	PluginClassHandler<CmsWindow, CompWindow>(window),
	window(window),
	cWindow(CompositeWindow::get(window)),
	gWindow(GLWindow::get(window)),
	isCms(false) {
    GLWindowInterface::setHandler(gWindow, true);

    updateMatch();
}

CmsWindow::~CmsWindow() {}

void CmsWindow::updateMatch() {
    CMS_SCREEN(screen);

    isCms = !cs->optionGetExcludeMatch().evaluate(window);

    cWindow->addDamage();
}

void CmsWindow::glDrawTexture(GLTexture *texture, const GLMatrix &transform,
	const GLWindowPaintAttrib &attrib, unsigned int mask) {
    CMS_SCREEN(screen);

    bool isDecoration = true;
    foreach (GLTexture *tex, gWindow->textures()) {
	if (texture == tex) {
	    isDecoration = false;
	    break;
	}
    }

    bool doCms = isDecoration ? cs->optionGetDecorations() : isCms;
    CmsLut *lut = 0;

    if (doCms) {
	CompRect borderRect = window->borderRect();
	foreach (CmsOutput& out, cs->cmsOutputs) {
	    if (borderRect.intersects(out.rect) && out.lut) {
		lut = out.lut;
		break;
	    }
	}
    }

    if (lut && doCms) {
	bool alpha = isDecoration || window->alpha();

	// FIXME: Figure out a better way to allocate the texture unit.
	int unit = 6; // > PrivateVertexBuffer::MAX_TEXTURES

	GL::activeTexture(GL_TEXTURE0 + unit);
	glBindTexture(GL_TEXTURE_3D, lut->texture_id);
	GL::activeTexture(GL_TEXTURE0);

	std::string fragment_shader = lut->getFragmentShader(alpha);
	gWindow->addShaders("cms", "", fragment_shader);
	gWindow->vertexBuffer()->addUniform("cms_lut", unit);

	gWindow->glDrawTexture(texture, transform, attrib, mask);
    } else {
	/* no cms */
	gWindow->glDrawTexture(texture, transform, attrib, mask);
    }
}


bool CmsPluginVTable::init() {
    if (!CompPlugin::checkPluginABI("core", CORE_ABIVERSION) ||
	!CompPlugin::checkPluginABI("composite", COMPIZ_COMPOSITE_ABI) ||
	!CompPlugin::checkPluginABI("opengl", COMPIZ_OPENGL_ABI))
	return false;

    return true;
}
